### PR TITLE
Support an arbitrary CSV delimiter

### PR DESCRIPTION
**Pull request by zhkvia**
Added a `format_csv_delimiter` setting for specifying an arbitrary CSV delimiter.
How it can be used:
- As a client argument:
    ```shell
    $ clickhouse-client --format_csv_delimiter=";" --query="INSERT INTO table FORMAT CSV"
    ```
- As a session setting:
    ```sql
    :) SET format_csv_delimiter=';'                                                                                                        │

    SET format_csv_delimiter = ';'

    Ok.

    :) SELECT * FROM table FORMAT CSV
    ```

I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

### CODE REVIEW

File: dbms/src/Interpreters/SettingsCommon.h [hunk 1]
```diff
@@ -706,4 +706,57 @@ struct SettingString
     }
 };
 
+struct SettingChar
+{
+private:
+    void checkStringIsACharacter(const String & x) const {
+        if (x.size() != 1)
+            throw Exception(std::string("A setting's value string has to be an exactly one character long"));
```
Comment (alexey-milovidov) [inline dbms/src/Interpreters/SettingsCommon.h]: It's not necessary to construct `std::string` explicitly.

Comment (alexey-milovidov) [inline dbms/src/Interpreters/SettingsCommon.h]: Missing ErrorCodes.

File: dbms/src/Interpreters/SettingsCommon.h [hunk 2]
```diff
@@ -706,4 +706,57 @@ struct SettingString
...
+    void set(const String & x)
+    {
+        checkStringIsACharacter(x);
+        value = x[0];
+        changed = true;
+    }
+
+    void set(const Field & x)
+    {
+        String s = safeGet<const String &>(x);
```
Comment (alexey-milovidov) [inline dbms/src/Interpreters/SettingsCommon.h]: We can use reference here to avoid copying.

File: docs/en/formats/csv.md [hunk 3]
```diff
@@ -2,9 +2,11 @@
...
 
-When parsing, all values can be parsed either with or without quotes. Both double and single quotes are supported. Rows can also be arranged without quotes. In this case, they are parsed up to a comma or line feed (CR or LF). In violation of the RFC, when parsing rows without quotes, the leading and trailing spaces and tabs are ignored. For the line feed, Unix (LF), Windows (CR LF) and Mac OS Classic (CR LF) are all supported.
+&ast;By default — `,`. See a [format_csv_delimiter](/docs/en/operations/settings/settings/#format_csv_delimiter) setting for additional info.
```
Comment (alexey-milovidov) [inline docs/en/formats/csv.md]: Are you sure that Markdown supports HTML entities?

**Comment by alexey-milovidov** [comment]
Almost everything is Ok, but tests are missing.
You may use simple functional tests (look at `dbms/tests/queries` directory).
Both input and output should be tested.
Please add test cases when unquoted string use comma and delimiter is a semicolon or something like this:
`abc,def;hello`

File: dbms/src/Interpreters/SettingsCommon.h (hunk 1 above)
Comment (ludv1x) [reply dbms/src/Interpreters/SettingsCommon.h]: Style. Braces `{}` should be in a separate new line.

File: docs/en/formats/csv.md (hunk 3 above)
Comment (luc1ph3r) [reply docs/en/formats/csv.md]: It seems logical since markdown is compiled into HTML. Also, `mkdocs` in the `docs/` directory compiles it correctly.

**Comment by amosbird** [comment]
Hmm, can we still call that a `CSV` format without comma? I assume a new format with some delimiter argument would be a better choice.

**Comment by alexey-milovidov** [comment]
> Hmm, can we still call that a CSV format without comma? I assume a new format with some delimiter argument would be a better choice.

Yes, it is controversial. In fact, many "CSV readers" are configurable in this way.
We may add a format with name 'DSV', but CSV is also Ok.

> And there are some use cases of multi-bytes delimiters such as data exported from Netezza.

Do they use multibyte delimiter by default?

**Comment by amosbird** [comment]
> Do they use multibyte delimiter by default?
Nope. The default is `|`.

**Comment by zhangxin0112** [comment]
does `--format_csv_delimiter=` support in v1.1.54385-stable?
Errors:
Bad arguments: unrecognised option '--format_csv_delimiter= '

**Comment by AntonSaykovsky** [comment]
> does --format_csv_delimiter= support in v1.1.54385-stable?
> Errors:
> Bad arguments: unrecognised option '--format_csv_delimiter= '

1.1.54390 version has the same error. =(((

**Pull request merged**